src/main/java/com/wultra/app/enrollmentserver/impl/service/MobileTokenService.java

```diff
@@ -0,0 +1,160 @@
+        final OperationDetailResponse operationDetailResponse = powerAuthClient.operationDetail(operationDetailRequest);
+        String status = operationDetailResponse.getStatus();
+        handleStatus(status);
+
+        final com.wultra.security.powerauth.client.model.request.OperationApproveRequest approveRequest = new com.wultra.security.powerauth.client.model.request.OperationApproveRequest();
+        approveRequest.setOperationId(operationId);
+        approveRequest.setData(data);
+        approveRequest.setUserId(userId);
+        approveRequest.setSignatureType(signatureFactors);
+        approveRequest.setApplicationId(applicationId);
+        final OperationUserActionResponse approveResponse = powerAuthClient.operationApprove(approveRequest);
+
+        final UserActionResult result = approveResponse.getResult();
+        if (result == UserActionResult.APPROVED) {
+            return new Response();
+        } else {
+            final OperationDetailResponse operation = approveResponse.getOperation();
+            status = operation.getStatus();
+            handleStatus(status);
+        }
+        throw new MobileTokenException("POWERAUTH_AUTH_FAIL", "Authentication failed");
+    }
+
+    public Response operationReject(
+            @NotNull String userId,
+            @NotNull Long applicationId,
+            @NotNull String operationId) throws MobileTokenException, PowerAuthClientException {
+        final OperationDetailRequest operationDetailRequest = new OperationDetailRequest();
+        operationDetailRequest.setOperationId(operationId);
+        final OperationDetailResponse operationDetailResponse = powerAuthClient.operationDetail(operationDetailRequest);
+        String status = operationDetailResponse.getStatus();
+        handleStatus(status);
+
+        final com.wultra.security.powerauth.client.model.request.OperationRejectRequest rejectRequest = new com.wultra.security.powerauth.client.model.request.OperationRejectRequest();
+        rejectRequest.setOperationId(operationId);
+        rejectRequest.setUserId(userId);
+        rejectRequest.setApplicationId(applicationId);
+
+        final OperationUserActionResponse rejectResponse = powerAuthClient.operationReject(rejectRequest);
+
+        final UserActionResult result = rejectResponse.getResult();
+        if (result == UserActionResult.REJECTED) {
+            return new Response();
+        } else {
+            final OperationDetailResponse operation = rejectResponse.getOperation();
+            status = operation.getStatus();
+            handleStatus(status);
+        }
+        throw new MobileTokenException("POWERAUTH_AUTH_FAIL", "Authentication failed");
+    }
+
+    private void handleStatus(String status) throws MobileTokenException {
+        switch (status) {
+            case "PENDING": {
+                // OK, this operation is still pending
+                break;
+            }
+            case "CANCELLED": {
```

Comment:
Typo: `CANCELLED` and `OPERATION_ALREADY_CANCELED`, choose one way. I prefer the variant with single L, which we use in Web Flow.